### ISSUE TITLE
RemoveFile does not remove

### ISSUE THREAD

**Issue by sudhiksk** [hierynomus/smbj#393]
I'm trying to move a file from remote src to remote dest. I first copy the file and then delete the source file. 

It works perfectly fine as an individual class, but when I have to use it in a bean class (.ear is deployed on liberty profile), Share.rm doesn't remove the file. It instead marks it as Status_Delete_Pending and gets deleted when I stop the server/JVM

**Comment by hierynomus** [comment]
Correct, you need to close the File handle before the delete is performed.

**Comment by sudhiksk** [comment]
Any help on this would be really appreciated. 

My app deployed as an ear file on unix server has to look for certain files in a windows share, delete those files once processed so that they are not picked up again. But, since these are marked for deletion, these are picked up again and they error our when try to open this file.

**Comment by ecki** [comment]
Can you actually delete the file locally? Maybe its just locked - Windows blocks deletes then. (I dont see a flaw in your  code - at least not in the part you have shown. When transfering the file you close it, right?)

**Comment by hierynomus** [comment]
If you close the Share, the file should be deleted. But I see that you're also not closing the `File` object in the `moveFile` method. Please ensure that you close all closeable resources.

**Comment by ecki** [comment]
yes your copy loop (writing to destination) is "transferring" the file. But anyway, no matter the termiinology, if you dont close the stream and the destFile it will be busy. (I am not sure why it stays busy when you close the client, however in all cases you need to close the inner resources. For the output stream it is particular important to close it and react to exceptions so you dont rename a broken result.

**Comment by sudhiksk** [comment]
I tried that as well but it doesn't work. I had removed it from above code since those are autocloseable. As soon as I stop the JVM(I don't know what resources it is holding), it deletes the files. 

I thought it be the EJBs that are having an object of my SMBJ Utility class, so I just did new SMBJUtility().moveFiles and it doesn't work as well.

I will keep looking, any other hints from you would be very helpful.

**Comment by sudhiksk** [comment]
I somewhere found that file handler was not being closed. This solved the issue and can be closed.

However, i've another issue which I can open another chain if required. Below lines of code that writes the file. It creates a file with current date time. How can I make the new file have the same properties as the source file.

Set<SMB2ShareAccess> s = new HashSet<>();
    s.add(SMB2ShareAccess.ALL.iterator().next());
    try(com.hierynomus.smbj.share.File srcFile = srcShare.openFile(srcPath, EnumSet.of(AccessMask.GENERIC_READ), null, s, SMB2CreateDisposition.FILE_OPEN, null);
            InputStream in = srcFile.getInputStream();
            com.hierynomus.smbj.share.File destFile = destShare.openFile(destPath, EnumSet.of(AccessMask.GENERIC_WRITE), null, SMB2ShareAccess.ALL, SMB2CreateDisposition.FILE_OVERWRITE_IF, null);
            OutputStream out = destFile.getOutputStream()){

            int bytesRead;
            byte[] buffer = new byte[1024];
            while((bytesRead = in.read(buffer)) != -1) {
                out.write(buffer, 0, bytesRead);
            }
            out.flush();

**Comment by ecki** [comment]
You really should use out.close() instead of out.flush() or use TWR.

To your question I think you can use share.setFileInformation(path, info) with info being BasicFileInformation which has setters for the timestamps.

BTW instead of streams the utility methods in smbj seem to prefer using the share.write() with a InputStreamChunkProvider instead.

**Comment by oleg30** [comment]
sudhiksk, tell us please what file handler did you close and how, for solving this issue...

**Comment by sudhiksk** [comment]
I had written a generic function getInputStream in my utility class and was being called from a different class. I was able to close the input stream in this class but filehandler was not closed.

**Comment by sandeepteerth** [comment]
I have written Java code to get the connection to share drive and reuse that connection for other operations. But for delete file I create separate connection and after file remove command I will close file handler and the connection to make sure file gets deleted immediately, but even after closing all file handler and closing the connection file is not getting deleted. When I close the connection which I had opened earlier for reuse, file is getting deleted . Can you please suggest ?

**Comment by ecki** [comment]
You need to close the write/upload or read (well actually all file access), not the delete command.

**Comment by sandeepteerth** [comment]
@ecki  pls have look at below code where I am closing all file handlers and connection.
SmbConfig config1 = SmbConfig.builder()
	            .withTimeout(timeout, TimeUnit.SECONDS) // Timeout sets Read, Write, and Transact timeouts (default is 60 seconds)
	            .withSoTimeout(socketTimeout, TimeUnit.SECONDS) // Socket Timeout (default is 0 seconds, blocks forever)
	            .withDfsEnabled(true)
	            .build();
		
		SMBClient client1 = new SMBClient(config1);
		Connection connection1 = client1.connect(host);
		
		AuthenticationContext ac1 = new AuthenticationContext(username, password.toCharArray(), domain);
		Session session1 = connection1.authenticate(ac1);
		  
		DiskShare share1 = (DiskShare) session1.connectShare(sharedlocation); 
		File smbFileRead1 = share1.openFile(filePath, EnumSet.of(AccessMask.GENERIC_READ), null, SMB2ShareAccess.ALL, SMB2CreateDisposition.FILE_OPEN, null);
		 File smbFileWrite1 = share1.openFile(
                 destPath,
                 EnumSet.of(AccessMask.GENERIC_WRITE),
                 EnumSet.of(FileAttributes.FILE_ATTRIBUTE_NORMAL),
                 EnumSet.of(SMB2ShareAccess.FILE_SHARE_WRITE),
                 (overwrite ? SMB2CreateDisposition.FILE_OVERWRITE_IF : SMB2CreateDisposition.FILE_CREATE),
                 EnumSet.noneOf(SMB2CreateOptions.class)
             );
		 
		 InputStream is = smbFileRead1.getInputStream(); 
		 OutputStream oStream = smbFileWrite1.getOutputStream();
		 
		 byte[] bytes = IOUtils.toByteArray(is);
		 oStream.write(bytes);
		 oStream.flush();
		 oStream.close();
		 smbFileWrite1.close();
		 is.close();
		 smbFileRead1.close();
		 
		 share1.rm(filePath);
		 share1.close();
		 session1.close();
		 connection1.close();
		 client1.close();

**Comment by ecki** [comment]
The close looks fine, maybe it's a bug server side. Did you try it with different servers? Maybe turn off malware scanners on the server, they always cause trouble. You would need to sleep for short time and retry in a loop.

**Comment by hierynomus** [comment]
Indeed, if in doubt, check the integration tests for code samples. You're missing a few try-finally blocks to ensure it always closes, or try-with-resources, but other than that this is how it should work.

**Comment by ecki** [comment]
Maybe one more thing, you enable dfs, are the two files actually on different machines? There have been some implicate reconnects going on in the background, maybe they are the reason. 

Can you try same source and target directory and maybe turn off DFS with a stand alone smb share?
